instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,382 @@
+// configs in the config{} block, there won't be an update request if their values are same as what have been read.
+func resourceKafkaTopicUpdate(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	log.Printf("[INFO] Updating the config of Kafka topic %s in %s.", topic, cluster_id)
+
+	// This is for when only replication_factor or partitions are changed, but we don't support changing them currently
+	if !d.HasChange("config") {
+		log.Printf("[INFO] Currently we only supprot updating topic's config. There are no changes in topic %s's config.", topic)
+		return nil
+	}
+	// This is for when there is no config{} block in the resource, or the whole config{} is removed from the resource.
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		return nil
+	}
+	client := meta.(*Config).Client
+
+	changedConfigMap, err := getChangedConfigMap(d)
+	if err != nil {
+		return err
+	}
+	if len(changedConfigMap) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		err = resourceKafkaTopicRead(d, meta)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var kafkaTopicConfigOptions KafkaTopicConfigOptions
+	err = mapstructure.Decode(changedConfigMap, &kafkaTopicConfigOptions)
+	if err != nil {
+		return err
+	}
+
+	updateKafkaTopicRequest := UpdateKafkaTopicRequest{
+		Config: &kafkaTopicConfigOptions,
+	}
+
+	jsonStr, err := json.Marshal(updateKafkaTopicRequest)
+	if err != nil {
+		return fmt.Errorf("[Error] Error creating kafka topic update request: %s", err)
+	}
+
+	err = client.UpdateKafkaTopic(cluster_id, topic, jsonStr)
+	if err != nil {
+		return fmt.Errorf("[Error] Error updating the config of Kafka topic %s: %s", topic, err)
+	}
+
+	//Reading updated configs
+	err = resourceKafkaTopicRead(d, meta)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] The configs of Kafka topic %s has been updated.", topic)
+	return nil
+}
+
+func getChangedConfigMap(d *schema.ResourceData) (map[string]interface{}, error) {
+	newConfig := d.Get("config").([]interface{})[0].(map[string]interface{})
+	changedConfigMap := make(map[string]interface{}, 0)
+	var kafkaTopicConfig KafkaTopicConfigOptions
+	decodedConfigMap := make(map[string]interface{}, 0) // This map stores all non-zero values and all the pointers
+
+	// Decode twice here because zero-value non-pointers need to be removed
+	err := mapstructure.Decode(newConfig, &kafkaTopicConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = mapstructure.Decode(kafkaTopicConfig, &decodedConfigMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// The following block is to get what configs are indeed changed
+	for k, v := range newConfig {
+		key := fmt.Sprintf("config.0.%s", k)
+		valueOfV := reflect.ValueOf(decodedConfigMap[k])
+		if d.HasChange(key) {
```

Comment:
I haven't played with this, but what happens if you just include the keys that hasChange? that is, without the additional processing using reflection

Reply:
What do you mean? reflection is not used here, it's for later usage.

Reply:
yeah I mean further processing using reflection and stuff, I think I missed the party a bit, your discussion with Arjun above seems to answered my question